pkg/instance/pool.go

```diff
@@ -0,0 +1,277 @@
+// Copyright 2024 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package instance
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/vm"
+)
+
+type Instance interface {
+	Index() int
+	Close()
+}
+
+type UpdateInfo func(cb func(info *Info))
+type Runner[T Instance] func(ctx context.Context, inst T, updInfo UpdateInfo) error
+type CreateInstance[T Instance] func(int) (T, error)
+
+// Pool[T] provides the functionality of a generic pool of instances.
+// The instance is assumed to boot, be controlled by one Runner and then be re-created.
+// The pool is assumed to have one default Runner (e.g. to be used for fuzzing), while a
+// dynamically controlled sub-pool might be reserved for the arbitrary Runners.
+type Pool[T Instance] struct {
+	BootErrors chan error
+
+	count      int
+	creator    CreateInstance[T]
+	defaultJob Runner[T]
+	jobs       chan Runner[T]
+
+	// The mutex serializes ReserveForRun() calls.
+	mu        sync.Mutex
+	instances []poolInstance[T]
+}
+
+func NewVMPool(pool *vm.Pool, def Runner[*vm.Instance]) *Pool[*vm.Instance] {
+	return NewPool(pool.Count(),
+		func(idx int) (*vm.Instance, error) {
+			return pool.Create(idx)
+		},
+		def)
+}
+
+func NewPool[T Instance](count int, creator CreateInstance[T], def Runner[T]) *Pool[T] {
+	instances := make([]poolInstance[T], count)
+	for i := 0; i < count; i++ {
+		instances[i].idx = i
+		instances[i].free(def)
+	}
+	return &Pool[T]{
+		BootErrors: make(chan error, 16),
+		count:      count,
+		creator:    creator,
+		defaultJob: def,
+		instances:  instances,
+		jobs:       make(chan Runner[T]),
+	}
+}
+
+func (p *Pool[T]) Loop(ctx context.Context) {
+	var wg sync.WaitGroup
+	wg.Add(p.count)
+	for i := 0; i < p.count; i++ {
+		i := i
+		go func() {
+			for ctx.Err() == nil {
+				p.runInstance(ctx, i)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+}
+
+func (p *Pool[T]) runInstance(ctx context.Context, idx int) {
+	obj := &p.instances[idx]
```

Comment:
In some cases it's called `inst`, in some cases `obj`. Let's call it consistently.